apps/src/templates/EnhancedSafeMarkdown.jsx

```diff
@@ -48,21 +48,77 @@ export const ExpandableImagesWrapper = connect(null, dispatch => ({
   },
 }))(UnconnectedExpandableImagesWrapper);
 
+// A callout can be added via markdown instructions like this:
+//   [play](#showcallout=play-sound-block)
+//
+// This code will replace any link that has #showcallout=[id]
+// with bold clickable text that, when clicked, will call the
+// provided showCallout function with the ID.
+export class ShowCalloutsWrapper extends React.Component {
+  static propTypes = {
+    showCallout: PropTypes.func.isRequired,
+    children: PropTypes.oneOfType([
+      PropTypes.arrayOf(PropTypes.node),
+      PropTypes.node,
+    ]).isRequired,
+  };
+
+  componentDidMount() {
+    this.postRenderHook();
+  }
+
+  componentDidUpdate(prevProps) {
+    // TODO: do we need to do any kind of cleanup here? Or otherwise do
+    // something more precise than calling the method again when we're
+    // responding to an update rather than an initial render?
+    this.postRenderHook();
+  }
+
+  postRenderHook() {
+    const thisNode = ReactDOM.findDOMNode(this);
+    this.renderShowCallouts(thisNode);
+  }
+
+  renderShowCallouts(node) {
+    const showCallouts = node.querySelectorAll('a[href]');
+    for (const showCallout of showCallouts) {
+      const regexp = /#showcallout=(.*)/;
+      const matches = showCallout.href.match(regexp);
+      if (matches?.length === 2) {
+        const calloutId = matches[1];
+        const textContent = showCallout.childNodes[0].textContent;
+        const newNode = document.createElement('span');
+        newNode.innerHTML = `<b style="cursor: pointer">${textContent}</b>`;
+        newNode.onclick = () => this.props.showCallout(calloutId);
+        showCallout.replaceWith(newNode);
```

Comment:
It's probably worth pointing out that manual DOM manipulation isn't the _ideal_ way to implement something like this.

Identifying the use of a particular syntax and rendering something particular as a result is what Remark is optimized for, and I think it would work very well here. We could add a remark plugin that recognizes the syntax you've defined and renders this styled `<b/>` tag as part of the underlying markdown rendering process; then all we'd need to do in post-render is add the click handlers. See for example how we implement `ExpandableImagesWrapper`; all it does is [search for `.expandable-image`](https://github.com/code-dot-org/code-dot-org/blob/05d945039f583d1380d562b25882e9ef71b41b23/apps/src/templates/utils/expandableImages.js#L19); it's the responsibility of the plugin to [actually add that class in the first place](https://github.com/code-dot-org/remark-plugins/blob/d8d480504fdf36350fa8b802b7883c3a2b7b0718/src/expandableImages.js#L27).

Of course, that's also quite a bit more convoluted to implement than the readable and approachable version you have here, so I'd say it's up for debate whether it's worth taking on that additional complexity just to avoid messing with React's rendered content.

Reply:
Thanks for this info!  I didn't know about the `remark-plugins` repo.  But I did get this error one time which was a cause for concern:

<img width="1373" alt="react-error" src="https://github.com/code-dot-org/code-dot-org/assets/2205926/47dffaea-63ce-4291-85ae-efcd42d3bbcb">

It sure looks like a side-effect of modifying the DOM after React generated it, and was a great incentive to follow your advice, which I've now prepped in these two PRs:

https://github.com/code-dot-org/code-dot-org/pull/55861
https://github.com/code-dot-org/remark-plugins/pull/50